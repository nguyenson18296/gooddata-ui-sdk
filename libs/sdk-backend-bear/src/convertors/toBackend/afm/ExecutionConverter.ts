// (C) 2007-2020 GoodData Corporation
import isEmpty from "lodash/isEmpty";
import { GdcExecuteAFM } from "@gooddata/api-model-bear";
import { convertFilters } from "./FilterConverter";
import { convertMeasure } from "./MeasureConverter";
import {
    dimensionsFindItem,
    dimensionTotals,
    IAttribute,
    IExecutionDefinition,
    ITotal,
    MeasureGroupIdentifier,
    totalIsNative,
} from "@gooddata/sdk-model";
import { toBearRef } from "../ObjRefConverter";

function convertAttribute(attribute: IAttribute, idx: number): GdcExecuteAFM.IAttribute {
    const alias = attribute.attribute.alias;
    const aliasProp = alias ? { alias } : {};
    return {
        displayForm: toBearRef(attribute.attribute.displayForm),
        localIdentifier: attribute.attribute.localIdentifier || `a${idx + 1}`,
        ...aliasProp,
    };
}

function convertAFM(def: IExecutionDefinition): GdcExecuteAFM.IAfm {
    const attributes: GdcExecuteAFM.IAttribute[] = def.attributes.map(convertAttribute);
    const attrProp = attributes.length ? { attributes } : {};

    const measures: GdcExecuteAFM.IMeasure[] = def.measures.map(convertMeasure);
    const measuresProp = measures.length ? { measures } : {};

    const filters: GdcExecuteAFM.CompatibilityFilter[] = convertFilters(def.filters);
    const filtersProp = filters.length ? { filters } : {};

    const nativeTotals = convertNativeTotals(def);
    const nativeTotalsProp = nativeTotals.length ? { nativeTotals } : {};

    return {
        ...measuresProp,
        ...attrProp,
        ...filtersProp,
        ...nativeTotalsProp,
    };
}

function convertNativeTotals(def: IExecutionDefinition): GdcExecuteAFM.INativeTotalItem[] {
    // first find all native totals defined across dimensions
    const nativeTotals = def.dimensions
        .map(dimensionTotals)
        .reduce((acc, totals) => acc.concat(...totals), [])
        .filter(totalIsNative);

    return nativeTotals.map((t) => {
        // then for each native total, look across buckets (if any) for an attribute that is specified in
        // the total definition
        const attributeInDims = dimensionsFindItem(def.dimensions, t.attributeIdentifier);

        if (!attributeInDims.length) {
            throw new Error(
                `Native total references attribute that is not in any dimension: ${t.attributeIdentifier}`,
            );
        } else if (attributeInDims.length > 1) {
            throw new Error(
                `Native total references attribute that is in multiple dimensions: ${t.attributeIdentifier}`,
            );
        }

        const attributeDim = attributeInDims[0];
        // now, knowing the dimension and index of the attribute.. roll up all attributes that are before it
        const rollupAttributes = attributeDim.dim.itemIdentifiers
            .slice(0, attributeDim.itemIdx)
            .filter((id) => id !== MeasureGroupIdentifier);

        // and create native total such, that it rolls up all those attributes
        return {
            measureIdentifier: t.measureIdentifier,
            attributeIdentifiers: rollupAttributes,
        };
    });
}

function convertTotals(totals: ITotal[] = []): GdcExecuteAFM.ITotalItem[] {
    return totals.map((t) => {
        return {
            type: t.type,
            attributeIdentifier: t.attributeIdentifier,
            measureIdentifier: t.measureIdentifier,
        };
    });
}

function convertDimensions(def: IExecutionDefinition): GdcExecuteAFM.IDimension[] {
    return def.dimensions.map((dim) => {
        const totals = convertTotals(dim.totals);
        const totalsProp = !isEmpty(totals) ? { totals } : {};

        return {
            itemIdentifiers: dim.itemIdentifiers,
            ...totalsProp,
        };
    });
}

export function convertResultSpec(def: IExecutionDefinition): GdcExecuteAFM.IResultSpec {
    const sortsProp = !isEmpty(def.sortBy) ? { sorts: def.sortBy } : {};
    const dims = convertDimensions(def);
    const dimsProp = !isEmpty(dims) ? { dimensions: dims } : {};

    return {
        ...sortsProp,
        ...dimsProp,
    };
}

// This has to be const function, otherwise api-extractor breaks...
/**
 * Converts execution definition to AFM Execution
 *
 * @param def - execution definition
 * @returns AFM Execution
 *
 * @internal
 */
export const toAfmExecution = (def: IExecutionDefinition): GdcExecuteAFM.IExecution => {
    return {
        execution: {
            afm: convertAFM(def),
            resultSpec: convertResultSpec(def),
        },
    };
};
