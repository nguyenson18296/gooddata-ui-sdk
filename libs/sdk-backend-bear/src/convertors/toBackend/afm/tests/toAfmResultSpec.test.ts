// (C) 2020 GoodData Corporation
import { toAfmExecution } from "../ExecutionConverter";
import { defWithAlias, defWithoutFilters } from "./InvalidInputs.fixture";
import { ReferenceLdm } from "@gooddata/reference-workspace";

import {
    emptyDef,
    newDefForItems,
    newPositiveAttributeFilter,
    defSetSorts,
    newAttributeSort,
    defSetDimensions,
    newDimension,
    newTotal,
    newTwoDimensional,
    MeasureGroupIdentifier,
    newDefForBuckets,
    newBucket,
    attributeLocalId,
    defWithFilters,
    newNegativeAttributeFilter,
} from "@gooddata/sdk-model";

const workspace = "test workspace";
const total = newTotal("nat", ReferenceLdm.Won, ReferenceLdm.Account.Name, "native total");

describe("converts execution definition to AFM Execution", () => {
    const Scenarios: Array<[string, any]> = [
        ["empty definition", emptyDef(workspace)],
        ["definition that has one attribute with alias and one attribute without localId", defWithAlias],
        ["definition that has no filter", defWithoutFilters],
        [
            "definition with filters",
            newDefForItems(
                workspace,
                [ReferenceLdm.Account.Name, ReferenceLdm.Activity.Subject, ReferenceLdm.Won],
                [newPositiveAttributeFilter(ReferenceLdm.Account.Name, ["myAccount"])],
            ),
        ],
        ["sorts", defSetSorts(emptyDef(workspace), [newAttributeSort(ReferenceLdm.Account.Name, "asc")])],
        ["dimensions", defSetDimensions(emptyDef(workspace), [newDimension(["localId1"])])],
        [
            "dimensions with native totals and attribute is in bucket",
            defSetDimensions(
                newDefForBuckets("test workspace", [
                    newBucket("mixedBucket1", ReferenceLdm.Activity.Default, ReferenceLdm.Won),
                    newBucket("measureBucket1", ReferenceLdm.WinRate),
                    newBucket("attributeBucket1", ReferenceLdm.Account.Name),
                ]),
                newTwoDimensional(
                    [attributeLocalId(ReferenceLdm.Account.Name)],
                    [MeasureGroupIdentifier, total],
                ),
            ),
        ],
    ];
    it.each(Scenarios)("AFM Execution with %s", (_desc, input) => {
        expect(toAfmExecution(input)).toMatchSnapshot();
    });

    it("throw error with dimensions with native totals but no attribute in bucket", () => {
        expect(() =>
            toAfmExecution(
                defSetDimensions(
                    emptyDef(workspace),
                    newTwoDimensional(
                        ["localId1"],
                        [
                            MeasureGroupIdentifier,
                            newTotal("nat", ReferenceLdm.Won, ReferenceLdm.Account.Name),
                        ],
                    ),
                ),
            ),
        ).toThrowErrorMatchingSnapshot();
    });

    it("should remove empty attribute filters and not cause RAIL-2083", () => {
        const emptyPositiveFilter = newPositiveAttributeFilter(ReferenceLdm.Product.Name, []);
        const emptyNegativeFilter = newNegativeAttributeFilter(ReferenceLdm.Product.Name, []);
        const positiveFilter = newPositiveAttributeFilter(ReferenceLdm.Product.Name, ["value 1"]);
        const negativeFilter = newNegativeAttributeFilter(ReferenceLdm.Product.Name, ["value 2"]);

        const def = defWithFilters(emptyDef("test"), [
            emptyPositiveFilter,
            emptyNegativeFilter,
            positiveFilter,
            negativeFilter,
        ]);
        const result = toAfmExecution(def);

        expect(result.execution.afm.filters).toMatchSnapshot();
    });
});
