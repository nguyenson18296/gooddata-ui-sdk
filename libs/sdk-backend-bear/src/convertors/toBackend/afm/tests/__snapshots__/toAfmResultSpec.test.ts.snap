// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts execution definition to AFM Execution AFM Execution with definition that has no filter 1`] = `
Object {
  "execution": Object {
    "afm": Object {},
    "resultSpec": Object {},
  },
}
`;

exports[`converts execution definition to AFM Execution AFM Execution with definition that has one attribute with alias and one attribute without localId 1`] = `
Object {
  "execution": Object {
    "afm": Object {
      "attributes": Array [
        Object {
          "alias": "alias",
          "displayForm": Object {
            "identifier": "attribute1",
          },
          "localIdentifier": "a_724874d1_attribute1",
        },
        Object {
          "displayForm": Object {
            "identifier": "attribute2",
          },
          "localIdentifier": "a_attribute2",
        },
      ],
    },
    "resultSpec": Object {},
  },
}
`;

exports[`converts execution definition to AFM Execution AFM Execution with definition with filters 1`] = `
Object {
  "execution": Object {
    "afm": Object {
      "attributes": Array [
        Object {
          "displayForm": Object {
            "identifier": "label.account.id.name",
          },
          "localIdentifier": "a_label.account.id.name",
        },
        Object {
          "displayForm": Object {
            "identifier": "label.activity.id.subject",
          },
          "localIdentifier": "a_label.activity.id.subject",
        },
      ],
      "filters": Array [
        Object {
          "positiveAttributeFilter": Object {
            "displayForm": Object {
              "identifier": "label.account.id.name",
            },
            "in": Object {
              "values": Array [
                "myAccount",
              ],
            },
          },
        },
      ],
      "measures": Array [
        Object {
          "definition": Object {
            "measure": Object {
              "item": Object {
                "identifier": "acugFHNJgsBy",
              },
            },
          },
          "localIdentifier": "m_acugFHNJgsBy",
        },
      ],
    },
    "resultSpec": Object {},
  },
}
`;

exports[`converts execution definition to AFM Execution AFM Execution with dimensions 1`] = `
Object {
  "execution": Object {
    "afm": Object {},
    "resultSpec": Object {
      "dimensions": Array [
        Object {
          "itemIdentifiers": Array [
            "localId1",
          ],
        },
      ],
    },
  },
}
`;

exports[`converts execution definition to AFM Execution AFM Execution with dimensions with native totals and attribute is in bucket 1`] = `
Object {
  "execution": Object {
    "afm": Object {
      "attributes": Array [
        Object {
          "displayForm": Object {
            "identifier": "label.activity.id",
          },
          "localIdentifier": "a_label.activity.id",
        },
        Object {
          "displayForm": Object {
            "identifier": "label.account.id.name",
          },
          "localIdentifier": "a_label.account.id.name",
        },
      ],
      "measures": Array [
        Object {
          "definition": Object {
            "measure": Object {
              "item": Object {
                "identifier": "acugFHNJgsBy",
              },
            },
          },
          "localIdentifier": "m_acugFHNJgsBy",
        },
        Object {
          "definition": Object {
            "measure": Object {
              "item": Object {
                "identifier": "abZgFKGPaGYM",
              },
            },
          },
          "localIdentifier": "m_abZgFKGPaGYM",
        },
      ],
      "nativeTotals": Array [
        Object {
          "attributeIdentifiers": Array [],
          "measureIdentifier": "m_acugFHNJgsBy",
        },
      ],
    },
    "resultSpec": Object {
      "dimensions": Array [
        Object {
          "itemIdentifiers": Array [
            "a_label.account.id.name",
          ],
        },
        Object {
          "itemIdentifiers": Array [
            "measureGroup",
          ],
          "totals": Array [
            Object {
              "attributeIdentifier": "a_label.account.id.name",
              "measureIdentifier": "m_acugFHNJgsBy",
              "type": "nat",
            },
          ],
        },
      ],
    },
  },
}
`;

exports[`converts execution definition to AFM Execution AFM Execution with empty definition 1`] = `
Object {
  "execution": Object {
    "afm": Object {},
    "resultSpec": Object {},
  },
}
`;

exports[`converts execution definition to AFM Execution AFM Execution with sorts 1`] = `
Object {
  "execution": Object {
    "afm": Object {},
    "resultSpec": Object {
      "sorts": Array [
        Object {
          "attributeSortItem": Object {
            "attributeIdentifier": "a_label.account.id.name",
            "direction": "asc",
          },
        },
      ],
    },
  },
}
`;

exports[`converts execution definition to AFM Execution should remove empty attribute filters and not cause RAIL-2083 1`] = `
Array [
  Object {
    "positiveAttributeFilter": Object {
      "displayForm": Object {
        "identifier": "label.product.id.name",
      },
      "in": Object {
        "values": Array [
          "value 1",
        ],
      },
    },
  },
  Object {
    "negativeAttributeFilter": Object {
      "displayForm": Object {
        "identifier": "label.product.id.name",
      },
      "notIn": Object {
        "values": Array [
          "value 2",
        ],
      },
    },
  },
]
`;

exports[`converts execution definition to AFM Execution throw error with dimensions with native totals but no attribute in bucket 1`] = `"Native total references attribute that is not in any dimension: a_label.account.id.name"`;
